tests/test_schemas.py

```diff
@@ -0,0 +1,303 @@
+        "icm_version": 1,
+        "image_layer_index": -1,
+    },
+}
+
+
+@pytest.mark.parametrize(
+    "data, schema, err_message",
+    [
+        # source_containers_user_params.json
+        ({}, "schemas/source_containers_user_params.json", "validating 'required' has failed"),
+        (SOURCE_CONTAINERS_USER_PARAMS_MINIMAL, "schemas/source_containers_user_params.json", None),
+        (SOURCE_CONTAINERS_USER_PARAMS_ALL_PROPERTIES,
+         "schemas/source_containers_user_params.json", None),
+        (SOURCE_CONTAINERS_USER_PARAMS_MISSING_USER,
+         "schemas/source_containers_user_params.json",
+         r"validating 'required' has failed \('user' is a required property\)"),
+        (SOURCE_CONTAINERS_USER_PARAMS_MISSING_BUILD_JSON_DIR,
+         "schemas/source_containers_user_params.json",
+         r"validating 'required' has failed \('build_json_dir' is a required property\)"),
+        (SOURCE_CONTAINERS_USER_PARAMS_ADDITIONAL_NONDEFINED_PROPERTY,
+         "schemas/source_containers_user_params.json", None),  # TODO should this fail?
+
+        # user_params.json
+        (USER_PARAMS_MINIMAL, "schemas/user_params.json", None),
+        (USER_PARAMS_MISSING_BUILD_JSON_DIR,
+         "schemas/user_params.json",
+         r"validating 'required' has failed \('build_json_dir' is a required property\)"),
+        (USER_PARAMS_MISSING_USER,
+         "schemas/user_params.json",
+         r"validating 'required' has failed \('user' is a required property\)"),
+        (USER_PARAMS_MISSING_GIT_URI,
+         "schemas/user_params.json",
+         r"validating 'required' has failed \('git_uri' is a required property\)"),
+        (USER_PARAMS_MISSING_GIT_REF,
+         "schemas/user_params.json",
+         r"validating 'required' has failed \('git_ref' is a required property\)"),
+        (USER_PARAMS_ADDITIONAL_NONDEFINED_PROPERTY,
+         "schemas/user_params.json", None),  # TODO should this fail?
+        ({}, "schemas/user_params.json", "validating 'required' has failed"),
+
+        # operator_csv_modifications.json
+        ({}, "schemas/operator_csv_modifications.json", "validating 'required' has failed"),
+        (OPERATOR_CSV_MODIFICATIONS_MINIMAL, "schemas/operator_csv_modifications.json", None),
+        (OPERATOR_CSV_MODIFICATIONS_MISSING_ORIGINAL_PULLSPEC,
+         "schemas/operator_csv_modifications.json",
+         r"validating 'required' has failed \('original' is a required property\)"),
+        (OPERATOR_CSV_MODIFICATIONS_MISSING_NEW_PULLSPEC, "schemas/operator_csv_modifications.json",
+         r"validating 'required' has failed \('new' is a required property\)"),
+        (OPERATOR_CSV_MODIFICATIONS_MISSING_PINNED_PULLSPEC,
+         "schemas/operator_csv_modifications.json",
+         r"validating 'required' has failed \('pinned' is a required property\)"),
+        ({"pullspec_replacements": []},
+         "schemas/operator_csv_modifications.json", None),  # TODO should this fail?
```

Comment:
Empty pullspec replacements would be valid if the original CSV had no pullspecs. Not sure how much sense that makes, but it looks like the code is ready to handle that, so this seems fine.